File: typescript/src/thresholds.ts
/**
 * Coherence thresholds - Generic bands for coherence state classification.
 *
 * (c) 2025 Anywave Creations
 * MIT License
 */

// Generic coherence thresholds (not application-specific)
/** High coherence threshold (85%) */
export const HIGH_COHERENCE: number = 0.85;

/** Medium coherence threshold (60%) */
export const MEDIUM_COHERENCE: number = 0.6;

/** Low coherence threshold (30%) */
export const LOW_COHERENCE: number = 0.3;

/** Minimum detectable coherence (10%) */
export const MINIMUM_COHERENCE: number = 0.1;

/** Definition of a coherence band */
export interface CoherenceBand {
  lower: number;  // Lower bound (inclusive)
  upper: number;  // Upper bound (exclusive, except for highest band)
  name: string;   // Human-readable name
}

/** Coherence level classification */
export const CoherenceLevel = {
  PEAK: { lower: HIGH_COHERENCE, upper: 1.01, name: "peak" },
  HIGH: { lower: MEDIUM_COHERENCE, upper: HIGH_COHERENCE, name: "high" },
  MEDIUM: { lower: LOW_COHERENCE, upper: MEDIUM_COHERENCE, name: "medium" },
  LOW: { lower: MINIMUM_COHERENCE, upper: LOW_COHERENCE, name: "low" },
  MINIMAL: { lower: 0.0, upper: MINIMUM_COHERENCE, name: "minimal" },
} as const satisfies Record<string, CoherenceBand>;

export type CoherenceLevelName = keyof typeof CoherenceLevel;

/**
 * Classify a coherence value into a level.
 * @param value Coherence value (0-1)
 * @returns The appropriate CoherenceLevel
 */
export function classifyCoherence(value: number): CoherenceBand {
  if (value < 0 || value > 1) {
    throw new Error("Coherence must be between 0 and 1");
  }

  for (const level of Object.values(CoherenceLevel)) {
    if (level.lower <= value && value < level.upper) {
      return level;
    }
  }

  // Handle edge case of exactly 1.0
  return CoherenceLevel.PEAK;
}

/**
 * Check if a value falls within a specific band.
 * @param band The coherence band to check against
 * @param value Coherence value to check
 * @returns True if value is in this band
 */
export function bandContains(band: CoherenceBand, value: number): boolean {
  return band.lower <= value && value < band.upper;
}

/**
 * Normalize a value to the 0-1 coherence range.
 * @param value Value to normalize
 * @param minVal Minimum of input range
 * @param maxVal Maximum of input range
 * @returns Normalized value between 0 and 1
 */
export function normalizeCoherence(
  value: number,
  minVal: number = 0.0,
  maxVal: number = 1.0
): number {
  if (maxVal <= minVal) {
    throw new Error("maxVal must be greater than minVal");
  }

  const normalized = (value - minVal) / (maxVal - minVal);
  return Math.max(0.0, Math.min(1.0, normalized));
}

/**
 * Calculate the change in coherence between two measurements.
 * @param current Current coherence value
 * @param previous Previous coherence value
 * @returns Delta (positive = increasing coherence)
 */
export function coherenceDelta(current: number, previous: number): number {
  return current - previous;
}

/**
 * Check if a series of coherence values is stable.
 * @param values List of coherence measurements
 * @param threshold Maximum allowed standard deviation
 * @returns True if coherence is stable
 */
export function isCoherenceStable(values: number[], threshold: number = 0.05): boolean {
  if (values.length < 2) return true;

  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  const variance = values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) / values.length;
  const stdDev = Math.sqrt(variance);

  return stdDev <= threshold;
}

File: rust/src/thresholds.rs
//! Coherence thresholds - Generic bands for coherence state classification.
//!
//! (c) 2025 Anywave Creations
//! MIT License

/// High coherence threshold (85%)
pub const HIGH_COHERENCE: f64 = 0.85;

/// Medium coherence threshold (60%)
pub const MEDIUM_COHERENCE: f64 = 0.6;

/// Low coherence threshold (30%)
pub const LOW_COHERENCE: f64 = 0.3;

/// Minimum detectable coherence (10%)
pub const MINIMUM_COHERENCE: f64 = 0.1;

/// Definition of a coherence band
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct CoherenceBand {
    /// Lower bound (inclusive)
    pub lower: f64,
    /// Upper bound (exclusive, except for highest band)
    pub upper: f64,
    /// Human-readable name
    pub name: &'static str,
}

impl CoherenceBand {
    /// Create a new coherence band
    pub const fn new(lower: f64, upper: f64, name: &'static str) -> Self {
        Self { lower, upper, name }
    }

    /// Check if a value falls within this band
    pub fn contains(&self, value: f64) -> bool {
        self.lower <= value && value < self.upper
    }
}

/// Coherence level classification
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CoherenceLevel {
    Peak,
    High,
    Medium,
    Low,
    Minimal,
}

impl CoherenceLevel {
    /// Get the band definition for this level
    pub const fn band(&self) -> CoherenceBand {
        match self {
            Self::Peak => CoherenceBand::new(HIGH_COHERENCE, 1.01, "peak"),
            Self::High => CoherenceBand::new(MEDIUM_COHERENCE, HIGH_COHERENCE, "high"),
            Self::Medium => CoherenceBand::new(LOW_COHERENCE, MEDIUM_COHERENCE, "medium"),
            Self::Low => CoherenceBand::new(MINIMUM_COHERENCE, LOW_COHERENCE, "low"),
            Self::Minimal => CoherenceBand::new(0.0, MINIMUM_COHERENCE, "minimal"),
        }
    }

    /// Get lower bound of this band
    pub const fn lower(&self) -> f64 {
        self.band().lower
    }

    /// Get upper bound of this band
    pub const fn upper(&self) -> f64 {
        self.band().upper
    }

    /// Classify a coherence value into a level
    ///
    /// # Arguments
    /// * `value` - Coherence value (0-1)
    ///
    /// # Returns
    /// The appropriate CoherenceLevel
    ///
    /// # Panics
    /// Panics if value is not between 0 and 1
    pub fn classify(value: f64) -> Self {
        if !(0.0..=1.0).contains(&value) {
            panic!("Coherence must be between 0 and 1");
        }

        let levels = [
            Self::Peak,
            Self::High,
            Self::Medium,
            Self::Low,
            Self::Minimal,
        ];

        for level in levels {
            let band = level.band();
            if band.lower <= value && value < band.upper {
                return level;
            }
        }

        // Handle edge case of exactly 1.0
        Self::Peak
    }

    /// Check if a value falls within this level's band
    pub fn contains(&self, value: f64) -> bool {
        self.band().contains(value)
    }
}

/// Normalize a value to the 0-1 coherence range.
///
/// # Arguments
/// * `value` - Value to normalize
/// * `min_val` - Minimum of input range
/// * `max_val` - Maximum of input range
///
/// # Returns
/// Normalized value between 0 and 1
///
/// # Panics
/// Panics if max_val <= min_val
pub fn normalize_coherence(value: f64, min_val: f64, max_val: f64) -> f64 {
    if max_val <= min_val {
        panic!("max_val must be greater than min_val");
    }

    let normalized = (value - min_val) / (max_val - min_val);
    normalized.clamp(0.0, 1.0)
}

/// Calculate the change in coherence between two measurements.
///
/// # Arguments
/// * `current` - Current coherence value
/// * `previous` - Previous coherence value
///
/// # Returns
/// Delta (positive = increasing coherence)
pub fn coherence_delta(current: f64, previous: f64) -> f64 {
    current - previous
}

/// Check if a series of coherence values is stable.
///
/// # Arguments
/// * `values` - Slice of coherence measurements
/// * `threshold` - Maximum allowed standard deviation
///
/// # Returns
/// True if coherence is stable
pub fn is_coherence_stable(values: &[f64], threshold: f64) -> bool {
    if values.len() < 2 {
        return true;
    }

    let mean = values.iter().sum::<f64>() / values.len() as f64;
    let variance = values.iter().map(|v| (v - mean).powi(2)).sum::<f64>() / values.len() as f64;
    let std_dev = variance.sqrt();

    std_dev <= threshold
}

/// Check if a series of coherence values is stable with default threshold of 0.05.
pub fn is_coherence_stable_default(values: &[f64]) -> bool {
    is_coherence_stable(values, 0.05)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_coherence_level_classify() {
        assert_eq!(CoherenceLevel::classify(0.9), CoherenceLevel::Peak);
        assert_eq!(CoherenceLevel::classify(0.7), CoherenceLevel::High);
        assert_eq!(CoherenceLevel::classify(0.4), CoherenceLevel::Medium);
        assert_eq!(CoherenceLevel::classify(0.2), CoherenceLevel::Low);
        assert_eq!(CoherenceLevel::classify(0.05), CoherenceLevel::Minimal);
    }

    #[test]
    fn test_normalize_coherence() {
        assert!((normalize_coherence(50.0, 0.0, 100.0) - 0.5).abs() < 1e-10);
        assert!((normalize_coherence(-10.0, 0.0, 100.0) - 0.0).abs() < 1e-10);
        assert!((normalize_coherence(150.0, 0.0, 100.0) - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_is_coherence_stable() {
        assert!(is_coherence_stable_default(&[0.5, 0.51, 0.49, 0.5]));
        assert!(!is_coherence_stable_default(&[0.1, 0.9, 0.1, 0.9]));
    }
}
